controllers/webhook/nim_account_webhook.go

```diff
@@ -0,0 +1,116 @@
+	client client.Client
+}
+
+func NewNimAccountValidator(client client.Client) admission.CustomValidator {
+	return &nimAccountValidator{client: client}
+}
+
+func (v *nimAccountValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	account := obj.(*nimv1.Account)
+
+	log := logf.FromContext(ctx).WithName("NIMAccountValidatingWebhook").
+		WithValues("namespace", account.Namespace, "account", account.Name)
+	log.Info("Validating NIM Account creation")
+
+	err = v.verifySingletonInNamespace(ctx, account)
+	if err != nil {
+		log.Error(err, "Rejecting NIM Account creation because checking singleton didn't pass")
+		return nil, err
+	}
+
+	err = v.verifyAPIKeySecret(ctx, nil, account)
+	if err != nil {
+		log.Error(err, "Rejecting NIM Account creation because checking API key Secret didn't pass")
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (v *nimAccountValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	account := newObj.(*nimv1.Account)
+
+	log := logf.FromContext(ctx).WithName("NIMAccountValidatingWebhook").
+		WithValues("namespace", account.Namespace, "account", account.Name)
+	log.Info("Validating NIM Account update")
+
+	err = v.verifyAPIKeySecret(ctx, oldObj.(*nimv1.Account), account)
+	if err != nil {
+		log.Error(err, "Rejecting NIM Account update because checking API key Secret didn't pass")
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (v *nimAccountValidator) ValidateDelete(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	// For deletion, nothing needs to be validated
+	return nil, nil
+}
+
+func (v *nimAccountValidator) verifyAPIKeySecret(ctx context.Context, oldAccount, newAccount *nimv1.Account) error {
+	if oldAccount == nil || !reflect.DeepEqual(oldAccount.Spec.APIKeySecret, newAccount.Spec.APIKeySecret) {
+		if newAccount.Spec.APIKeySecret.Name == "" {
+			msg := "name of the Secret for the API key is invalid"
+			return field.Invalid(field.NewPath("spec").Child("apiKeySecret").Child("name"), newAccount.Spec.APIKeySecret.Name, msg)
+		}
+		name := newAccount.Spec.APIKeySecret.Name
+		namespace := newAccount.Spec.APIKeySecret.Namespace
+		if namespace == "" {
+			namespace = newAccount.Namespace
+		}
+
+		err := v.client.Get(ctx,
+			types.NamespacedName{
+				Name:      name,
+				Namespace: namespace,
+			},
+			&corev1.Secret{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				msg := "Secret for the API key cannot be found"
+				return field.Invalid(field.NewPath("spec").Child("apiKeySecret").Child("name"), name, msg)
+			} else {
+				return fmt.Errorf("failed to verify the API key Secret with err: %s", err.Error())
+			}
+		}
+	}
+	return nil
+}
+
+func (v *nimAccountValidator) verifySingletonInNamespace(ctx context.Context, account *nimv1.Account) error {
+	accountList := nimv1.AccountList{}
+	err := v.client.List(ctx,
```

Comment:
Can we do a get instead of a list call?

Reply:
I am not sure how to do a GET without knowing the name and namespace of the object.

Could you please be more specific, and maybe give me an example? Thanks!

Reply:
Oh I see. Then you'll need LIST.

Optional: It might be more efficient if you have any label selectors associated so that you don't have to scan the entire namespace.

Reply:
Label selectors won't help; we're trying to prevent two Accounts from existing in the same namespace because, for each Account, we're creating a _ConfigMap_ a _Template_ and a _Secret_ with a static naming convention. We don't want the two Accounts to overwrite each other's resources.

If we depend on labels, users can potentially create an Account without a label to work around our webhook.

See https://github.com/opendatahub-io/odh-model-controller/pull/289#discussion_r1835717365.

Reply:
That makes sense. Thanks for clarifying. Then this change looks good.